astrodata/fits.py

```diff
@@ -326,6 +329,130 @@ def normalize_indices(slc, nitems):
+                        .format(extver or ''))
+
+    # ASDF can only dump YAML to a binary file object, so do that and read
+    # the contents back from it for storage in a FITS extension:
+    with BytesIO() as fd:
+        with af:
+            # Generate the YAML, dumping any binary arrays as text:
+            af.write_to(fd, all_array_storage='inline')
+        fd.seek(0)
+        wcsbuf = fd.read()
+
+    # Convert the bytes to readable lines of text for storage (falling back to
+    # saving as binary in the unexpected event that this is not possible):
+    try:
+        wcsbuf = wcsbuf.decode('ascii').splitlines()
+    except UnicodeDecodeError:
+        # This should not happen, but if the ASDF contains binary data in
+        # spite of the 'inline' option above, we have to dump the bytes to
+        # a non-human-readable binary table rather than an ASCII one:
+        LOGGER.warning("Could not convert WCS {} ASDF to ASCII; saving table "
+                       "as binary".format(extver or ''))
+        hduclass = BinTableHDU
+        fmt = 'B'
+        wcsbuf = np.frombuffer(wcsbuf, dtype=np.uint8)
+    else:
+        hduclass = TableHDU
+        fmt = 'A{0}'.format(max(len(line) for line in wcsbuf))
+
+    # Construct the FITS table extension:
+    col = Column(name='gWCS', format=fmt, array=wcsbuf,
+                 ascii= hduclass is TableHDU)
+    hdu = hduclass.from_columns([col], name='WCS', ver=extver)
+
+    return hdu
+
+def asdftablehdu_to_wcs(hdu):
+    """
+    Recreate a gWCS object from its serialization in a FITS table extension.
+
+    Returns None (issuing a warning) if the extension cannot be parsed, so
+    the rest of the file can still be read.
+    """
+
+    import asdf
+
+    ver = hdu.header.get('EXTVER', -1)
+
+    if isinstance(hdu, (TableHDU, BinTableHDU)):
+        try:
+            colarr = hdu.data['gWCS']
+        except KeyError:
+            LOGGER.warning("Ignoring 'WCS' extension {} with no 'gWCS' table "
+                           "column".format(ver))
+            return
+
+        # If this table column contains text strings as expected, join the rows
+        # as separate lines of a string buffer and encode the resulting YAML as
+        # bytes that ASDF can parse. If AstroData has produced another format,
+        # it will be a binary dump due to the unexpected presence of non-ASCII
+        # data, in which case we just extract unmodified bytes from the table.
+        if colarr.dtype.kind in ('U', 'S'):
+            sep = os.linesep
+            # Not sure whether io.fits ever produces 'S' on Py 3, but just in
+            # case: join lines as str & avoid a TypeError with unicode linesep;
+            # could also use astype('U') but it assumes an encoding implicitly.
+            if colarr.dtype.kind == 'S' and not isinstance(sep, bytes):
+                colarr = np.vectorize(lambda b: b.decode('ascii'))(colarr)
```

Comment:
`np.chararray.decode` provides a simpler way for this.

Reply:
Done. There seems to be some problem with using `np.vectorize` on a `chararray` anyway, which puzzles me a bit since I tested this stuff, but I might have tested it a bit out of context without knowing the `io.fits` option to get a byte array.